File: src/data/tutorials/030-lru-cache.js
/**
 * Tutorial content for #030 — LRU Cache
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Design a data structure that follows the Least Recently Used (LRU) cache eviction policy. Support two O(1) operations: get(key) — return the value if the key exists, otherwise -1; and put(key, value) — insert or update the key. When the cache reaches its capacity and a new key must be inserted, evict the least recently used item first.`,
    example: `capacity = 2\nput(1, 1) → cache: {1=1}\nput(2, 2) → cache: {1=1, 2=2}\nget(1)    → returns 1, cache order: {2=2, 1=1} (1 is most recent)\nput(3, 3) → evicts key 2 (least recent), cache: {1=1, 3=3}\nget(2)    → returns -1 (was evicted)\n✅ Answer: -1`,
    keyInsight: `O(1) get and put require fast lookup (hash map) AND fast reordering (doubly-linked list). Together, the map gives instant node access and the list lets you move any node to the "most recent" position in O(1) by adjusting just its neighbours' pointers.`,
  },

  approaches: {
    LinkedHashMap: {
      intuition: `Many languages provide an ordered map that maintains insertion/access order — Java's LinkedHashMap, Python's OrderedDict, or JavaScript's Map (which preserves insertion order). You can use this built-in structure to implement LRU: on every access, move the key to the end ("most recent"). On eviction, remove the key at the front ("least recent").`,
      steps: [
        `Store capacity and create an ordered map (LinkedHashMap / OrderedDict / Map).`,
        `get(key): if the key is missing return -1. Otherwise, delete it and re-insert it (moving it to the "most recent" end), then return the value.`,
        `put(key, value): if key already exists, delete it so it can be re-inserted at the end.`,
        `If the map is now at capacity, delete the first (oldest) entry.`,
        `Insert (key, value) at the end of the map.`,
      ],
      example: `capacity = 2\n\nput(1,1): map = {1:1}\nput(2,2): map = {1:1, 2:2}\nget(1):   delete 1, re-insert → map = {2:2, 1:1}, return 1\nput(3,3): at capacity, evict first key (2) → map = {1:1}, insert 3 → map = {1:1, 3:3}\nget(2):   key not found → return -1\n✅ Answer: -1`,
      keyInsight: `O(1) average for all operations. Extremely concise — the ordered map does all the work. In Python this is a 10-line solution with OrderedDict.`,
    },

    'Doubly-Linked List + HashMap': {
      intuition: `Build the data structure from scratch. A hash map gives O(1) lookup: key → node. A doubly-linked list maintains the usage order: the head (or a sentinel head) holds the least-recently-used item, the tail holds the most-recently-used. Moving a node to the tail on every access or insert, and removing from the head on eviction, are both O(1) pointer operations because each node knows its prev and next.`,
      steps: [
        `Create a doubly-linked list with sentinel head and tail nodes (dummy nodes simplify edge cases).`,
        `Create a hash map: key → list node.`,
        `get(key): if key not in map return -1. Otherwise move that node to just before the tail (mark it "most recent"), then return its value.`,
        `put(key, value): if key exists, update its value and move it to the tail. If key is new, create a new node and insert it just before the tail; add it to the map.`,
        `If size now exceeds capacity, remove the node just after the sentinel head (the LRU node) and delete its key from the map.`,
        `Helper functions addToTail(node) and removeNode(node) manipulate the four surrounding pointers each.`,
      ],
      example: `capacity = 2\nSentinels: HEAD ↔ TAIL\n\nput(1,1): HEAD ↔ [1:1] ↔ TAIL, map={1:node1}\nput(2,2): HEAD ↔ [1:1] ↔ [2:2] ↔ TAIL, map={1,2}\nget(1):   move node1 to tail → HEAD ↔ [2:2] ↔ [1:1] ↔ TAIL, return 1\nput(3,3): size=3 > cap=2, evict LRU=node after HEAD=node2\n          HEAD ↔ [1:1] ↔ TAIL, map={1}\n          insert node3 → HEAD ↔ [1:1] ↔ [3:3] ↔ TAIL, map={1,3}\nget(2):   not in map → -1\n✅ Answer: -1`,
      keyInsight: `O(1) time for both get and put. O(capacity) space. This is the canonical interview solution — understanding how to maintain a doubly-linked list alongside a hash map is a fundamental systems design pattern.`,
    },

    'Timestamp (Not O(1))': {
      intuition: `The simplest possible LRU eviction strategy: store the last-access timestamp alongside every cache entry. On each get or put, record the current timestamp for that key. When capacity is exceeded and eviction is needed, scan all entries to find the one with the smallest timestamp — that is the least recently used. Straightforward to understand and implement, but the O(n) eviction scan violates the O(1) requirement.`,
      steps: [
        `Maintain two maps: values (key → value) and times (key → last-access timestamp). Keep a global counter ts.`,
        `get(key): if key not in values return -1. Otherwise update times[key] = ts++, return values[key].`,
        `put(key, value): if key exists, update value and times[key] = ts++.`,
        `If key is new and at capacity: find lruKey = key with minimum value in times map (O(n) scan). Remove it from both maps.`,
        `Insert values[key] = value, times[key] = ts++.`,
      ],
      example: `capacity=2, ts starts at 0\n\nput(1,1): values={1:1}, times={1:0}, ts=1\nput(2,2): values={1:1,2:2}, times={1:0,2:1}, ts=2\nget(1):   times={1:2,2:1}, return 1, ts=3\nput(3,3): at capacity. Scan times: min is key 2 (ts=1). Evict 2.\n          values={1:1,3:3}, times={1:2,3:3}, ts=4\nget(2):   key 2 not in values → return -1\n✅ Answer: -1 (correct result, but eviction was O(n))`,
      keyInsight: `O(1) get/put amortised for non-eviction operations, but O(n) when eviction is triggered. Does NOT satisfy the O(1) requirement. Presented as a baseline showing why a proper ordered structure (linked list or LinkedHashMap) is necessary.`,
    },
  },
}

File: public/cpp/030-lru-cache.cpp
/*
 * LeetCode Problem #146: LRU Cache
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/lru-cache/
 * 
 * Problem Statement:
 * Design a data structure that follows the constraints of a Least Recently Used (LRU) cache.
 * 
 * Implement the LRUCache class:
 * - LRUCache(int capacity) Initialize the LRU cache with positive size capacity.
 * - int get(int key) Return the value of the key if the key exists, otherwise return -1.
 * - void put(int key, int value) Update the value of the key if the key exists.
 *   Otherwise, add the key-value pair to the cache. If the number of keys exceeds
 *   the capacity from this operation, evict the least recently used key.
 * 
 * The functions get and put must each run in O(1) average time complexity.
 */

#include <unordered_map>
#include <list>
#include <iostream>
using namespace std;

// ==================== APPROACH 1: HashMap + Doubly Linked List (Optimal) ====================
/*
 * Algorithm:
 * - Use doubly linked list to maintain order (most recent at head)
 * - Use hash map for O(1) access to nodes
 * - get(): Move accessed node to head
 * - put(): Add to head, evict tail if capacity exceeded
 * 
 * Time Complexity: O(1) for both get and put
 * Space Complexity: O(capacity)
 * 
 * When to use: This is the OPTIMAL solution
 * 
 * Key Insight:
 * - Doubly linked list allows O(1) removal and insertion
 * - Hash map provides O(1) lookup
 * - Head = most recently used, Tail = least recently used
 */

// Custom doubly linked list node
struct Node {
    int key;
    int value;
    Node* prev;
    Node* next;
    
    Node(int k, int v) : key(k), value(v), prev(nullptr), next(nullptr) {}
};

class LRUCache_Optimal {
private:
    int capacity;
    unordered_map<int, Node*> cache;  // key -> node pointer
    Node* head;  // Most recently used
    Node* tail;  // Least recently used
    
    // Add node right after head (most recent position)
    void addNode(Node* node) {
        node->next = head->next;
        node->prev = head;
        head->next->prev = node;
        head->next = node;
    }
    
    // Remove node from its current position
    void removeNode(Node* node) {
        node->prev->next = node->next;
        node->next->prev = node->prev;
    }
    
    // Move node to head (mark as most recently used)
    void moveToHead(Node* node) {
        removeNode(node);
        addNode(node);
    }
    
    // Remove least recently used node (tail)
    Node* removeTail() {
        Node* node = tail->prev;
        removeNode(node);
        return node;
    }
    
public:
    LRUCache_Optimal(int capacity) : capacity(capacity) {
        // Create dummy head and tail
        head = new Node(0, 0);
        tail = new Node(0, 0);
        head->next = tail;
        tail->prev = head;
    }
    
    int get(int key) {
        if (cache.find(key) == cache.end()) {
            return -1;
        }
        
        Node* node = cache[key];
        moveToHead(node);  // Mark as recently used
        return node->value;
    }
    
    void put(int key, int value) {
        if (cache.find(key) != cache.end()) {
            // Key exists, update value and move to head
            Node* node = cache[key];
            node->value = value;
            moveToHead(node);
        } else {
            // New key
            Node* newNode = new Node(key, value);
            cache[key] = newNode;
            addNode(newNode);
            
            if (cache.size() > capacity) {
                // Evict LRU
                Node* lru = removeTail();
                cache.erase(lru->key);
                delete lru;
            }
        }
    }
    
    ~LRUCache_Optimal() {
        // Clean up all nodes
        Node* current = head;
        while (current != nullptr) {
            Node* next = current->next;
            delete current;
            current = next;
        }
    }
};

// ==================== APPROACH 2: Using STL list and unordered_map ====================
/*
 * Algorithm:
 * - Use std::list for doubly linked list
 * - Use unordered_map to store key -> list iterator
 * - Leverage STL's splice for O(1) move operations
 * 
 * Time Complexity: O(1) for both get and put
 * Space Complexity: O(capacity)
 * 
 * When to use: When you want cleaner code using STL
 * 
 * Key Insight:
 * - STL list provides splice() for O(1) element movement
 * - Store iterators in map for O(1) access
 */
class LRUCache_STL {
private:
    int capacity;
    list<pair<int, int>> cache;  // list of (key, value) pairs
    unordered_map<int, list<pair<int, int>>::iterator> map;  // key -> iterator
    
public:
    LRUCache_STL(int capacity) : capacity(capacity) {}
    
    int get(int key) {
        if (map.find(key) == map.end()) {
            return -1;
        }
        
        // Move to front (most recently used)
        cache.splice(cache.begin(), cache, map[key]);
        return map[key]->second;
    }
    
    void put(int key, int value) {
        if (map.find(key) != map.end()) {
            // Update existing key
            cache.splice(cache.begin(), cache, map[key]);
            map[key]->second = value;
        } else {
            // Add new key
            if (cache.size() == capacity) {
                // Evict LRU (back of list)
                int lruKey = cache.back().first;
                cache.pop_back();
                map.erase(lruKey);
            }
            
            cache.push_front({key, value});
            map[key] = cache.begin();
        }
    }
};

// ==================== APPROACH 3: Ordered Map (Not O(1)) ====================
/*
 * Algorithm:
 * - Use ordered map with timestamp as key
 * - Track access time for each key
 * - Find and remove oldest on eviction
 * 
 * Time Complexity: O(log n) for operations
 * Space Complexity: O(capacity)
 * 
 * When to use: NOT recommended - doesn't meet O(1) requirement
 * 
 * Note: This approach is shown for comparison but doesn't meet
 * the problem's O(1) requirement
 */
class LRUCache_OrderedMap {
private:
    int capacity;
    int timestamp;
    unordered_map<int, int> values;  // key -> value
    unordered_map<int, int> times;   // key -> last access time
    
public:
    LRUCache_OrderedMap(int capacity) : capacity(capacity), timestamp(0) {}
    
    int get(int key) {
        if (values.find(key) == values.end()) {
            return -1;
        }
        
        times[key] = timestamp++;
        return values[key];
    }
    
    void put(int key, int value) {
        if (values.find(key) != values.end()) {
            values[key] = value;
            times[key] = timestamp++;
        } else {
            if (values.size() == capacity) {
                // Find LRU key (O(n) operation)
                int lruKey = -1;
                int minTime = INT_MAX;
                for (auto& p : times) {
                    if (p.second < minTime) {
                        minTime = p.second;
                        lruKey = p.first;
                    }
                }
                values.erase(lruKey);
                times.erase(lruKey);
            }
            
            values[key] = value;
            times[key] = timestamp++;
        }
    }
};

// Use optimal implementation by default
typedef LRUCache_Optimal LRUCache;

// ==================== TEST CASES ====================
void runTests() {
    cout << "Test 1: Basic operations" << endl;
    LRUCache cache1(2);
    cache1.put(1, 1);
    cache1.put(2, 2);
    cout << "get(1): " << cache1.get(1) << endl;  // Expected: 1
    cache1.put(3, 3);  // Evicts key 2
    cout << "get(2): " << cache1.get(2) << endl;  // Expected: -1
    cache1.put(4, 4);  // Evicts key 1
    cout << "get(1): " << cache1.get(1) << endl;  // Expected: -1
    cout << "get(3): " << cache1.get(3) << endl;  // Expected: 3
    cout << "get(4): " << cache1.get(4) << endl;  // Expected: 4
    
    cout << "\nTest 2: Update existing key" << endl;
    LRUCache cache2(2);
    cache2.put(1, 1);
    cache2.put(2, 2);
    cache2.put(1, 10);  // Update key 1
    cout << "get(1): " << cache2.get(1) << endl;  // Expected: 10
    cache2.put(3, 3);  // Evicts key 2
    cout << "get(2): " << cache2.get(2) << endl;  // Expected: -1
    
    cout << "\nTest 3: Single capacity" << endl;
    LRUCache cache3(1);
    cache3.put(1, 1);
    cout << "get(1): " << cache3.get(1) << endl;  // Expected: 1
    cache3.put(2, 2);  // Evicts key 1
    cout << "get(1): " << cache3.get(1) << endl;  // Expected: -1
    cout << "get(2): " << cache3.get(2) << endl;  // Expected: 2
    
    cout << "\nTest 4: Access pattern" << endl;
    LRUCache cache4(3);
    cache4.put(1, 1);
    cache4.put(2, 2);
    cache4.put(3, 3);
    cout << "get(1): " << cache4.get(1) << endl;  // Expected: 1 (1 becomes most recent)
    cache4.put(4, 4);  // Evicts key 2 (LRU)
    cout << "get(2): " << cache4.get(2) << endl;  // Expected: -1
    cout << "get(1): " << cache4.get(1) << endl;  // Expected: 1
    cout << "get(3): " << cache4.get(3) << endl;  // Expected: 3
    cout << "get(4): " << cache4.get(4) << endl;  // Expected: 4
}

int main() {
    runTests();
    return 0;
}

/*
 * COMPARISON OF APPROACHES:
 * 
 * 1. HashMap + Custom Doubly Linked List (RECOMMENDED):
 *    Time: O(1) for get and put
 *    Space: O(capacity)
 *    Most control, optimal performance
 * 
 * 2. HashMap + STL list:
 *    Time: O(1) for get and put
 *    Space: O(capacity)
 *    Cleaner code, uses STL
 * 
 * 3. Ordered Map with Timestamp:
 *    Time: O(n) for eviction
 *    Space: O(capacity)
 *    Does NOT meet O(1) requirement
 * 
 * INTERVIEW TIPS:
 * - Explain why both hash map and linked list are needed
 * - Hash map provides O(1) lookup
 * - Doubly linked list provides O(1) insertion/deletion
 * - Draw the data structure showing both components
 * - Explain eviction policy clearly
 * - Discuss dummy head/tail nodes to simplify edge cases
 * - Mention STL alternative if time permits
 * - Walk through example showing cache state changes
 * 
 * KEY INSIGHTS:
 * - Need O(1) for both access and eviction
 * - Hash map alone can't maintain order
 * - Linked list alone can't provide O(1) lookup
 * - Combination gives best of both worlds
 * - Doubly linked list needed for O(1) removal
 * - Dummy nodes simplify boundary conditions
 * 
 * STEP-BY-STEP for capacity=2:
 * 
 * put(1,1): [1] (1 is MRU)
 * put(2,2): [2,1] (2 is MRU)
 * get(1):   [1,2] (1 becomes MRU, returns 1)
 * put(3,3): [3,1] (evict 2, 3 is MRU)
 * get(2):   [3,1] (returns -1, not found)
 * put(4,4): [4,3] (evict 1, 4 is MRU)
 * 
 * COMMON MISTAKES:
 * - Using singly linked list (can't remove in O(1))
 * - Not updating order on get() operations
 * - Forgetting to remove from hash map on eviction
 * - Memory leaks (not deleting evicted nodes)
 * - Not handling capacity = 1 edge case
 * - Incorrect eviction (removing wrong element)
 * - Not using dummy nodes (complex edge case handling)
 * 
 * FOLLOW-UP QUESTIONS:
 * - Implement LFU (Least Frequently Used) cache? (LeetCode #460)
 * - How to make it thread-safe? (Add locks/mutexes)
 * - What if we need to persist to disk? (Add serialization)
 * - How to handle cache invalidation? (Add TTL/expiration)
 * - What about distributed caching? (Redis, Memcached)
 * 
 * RELATED PROBLEMS:
 * - LFU Cache (LeetCode #460)
 * - Design HashMap (LeetCode #706)
 * - Design HashSet (LeetCode #705)
 * - All O(1) Data Structure (LeetCode #432)
 * - Design In-Memory File System (LeetCode #588)
 * 
 * VISUALIZATION of data structure:
 * 
 * HashMap:        Doubly Linked List:
 * key -> node     head <-> node1 <-> node2 <-> tail
 *                 (MRU)                        (LRU)
 * 
 * After get(key):
 * - Find node via HashMap: O(1)
 * - Remove from current position: O(1)
 * - Add to head: O(1)
 * 
 * After put(key, value) when full:
 * - Remove tail node: O(1)
 * - Remove from HashMap: O(1)
 * - Add new node to head: O(1)
 * - Add to HashMap: O(1)
 */

// Made with Bob
